src/PhenX.EntityFrameworkCore.BulkInsert/Metadata/PropertyAccessor.cs

```diff
@@ -1,66 +1,53 @@
 
-    private static Getter<object, object?> CreateInternalUntypedGetter<TSource, TValue>(PropertyInfo propertyInfo)
-    {
-        var getter = CreateGetter<TSource, TValue>(propertyInfo);
+        var instanceParam = Expression.Parameter(typeof(object), "instance");
 
-        return source => getter((TSource)source!);
-    }
+        // Convert object to the declaring type
+        Expression typedInstance = propertyInfo.DeclaringType!.IsValueType
+            ? Expression.Unbox(instanceParam, propertyInfo.DeclaringType)
+            : Expression.Convert(instanceParam, propertyInfo.DeclaringType);
 
-    public static Getter<TSource, TValue> CreateGetter<TSource, TValue>(PropertyInfo propertyInfo)
-    {
-        if (!propertyInfo.CanRead)
-        {
-            return x => throw new NotSupportedException();
-        }
+        // Call Getter
+        Expression getterCall = Expression.Call(typedInstance, getMethod);
 
-        var bakingField =
-            propertyInfo.DeclaringType!.GetField($"<{propertyInfo.Name}>k__BackingField",
-                BindingFlags.NonPublic |
-                BindingFlags.Instance);
+        var propertyType = propertyInfo.PropertyType;
 
-        var propertyGetMethod = propertyInfo.GetGetMethod()!;
-
-        var getMethod = new DynamicMethod(propertyGetMethod.Name, typeof(TValue), [typeof(TSource)], true);
-        var getGenerator = getMethod.GetILGenerator();
-
-        // Load this to stack.
-        getGenerator.Emit(OpCodes.Ldarg_0);
-
-        if (bakingField != null && !propertyGetMethod.IsVirtual)
-        {
-            // Get field directly.
-            getGenerator.Emit(OpCodes.Ldfld, bakingField);
-        }
-        else if (propertyGetMethod.IsVirtual)
-        {
-            // Call the virtual property.
-            getGenerator.Emit(OpCodes.Callvirt, propertyGetMethod);
-        }
-        else
+        // If the converter is provided, we call it
+        if (converter != null)
         {
-            // Call the non virtual property.
-            getGenerator.Emit(OpCodes.Call, propertyGetMethod);
+            // Validate the converter input type matches property type
+            var convIn = converter.Parameters[0].Type;
+            if (!convIn.IsAssignableFrom(propertyType) && !propertyType.IsAssignableFrom(convIn))
+            {
+                throw new ArgumentException($"Converter input must be assignable from property type ({propertyType} -> {convIn})");
+            }
+
+            // If property type != converter param, convert
+            var converterInput = getterCall;
+            if (convIn != propertyType)
+            {
+                converterInput = Expression.Convert(getterCall, convIn);
+            }
+
+            getterCall = Expression.Invoke(converter, converterInput);
+
+            propertyType = getterCall.Type;
         }
 
-        getGenerator.Emit(OpCodes.Ret);
+        var finalExpr = propertyType.IsValueType
```

Comment:
I would just write Expression, I don't see a good reason to have a shortcut.

Reply:
What do you mean?